### ISSUE TITLE
Associate files and projects

### ISSUE THREAD

**Issue by privat** [nitlang/nit#1250]
TL;DR: In the specification, what are the rules to: 1. detect that a directory is the root directory of a project? 2. find all the modules of a project.

Currently, file are associated to groups and projects with various heuristics.

* if a file has the same base-name that the directory, then the directory is a group: all nit file in the directory belong to the group
* if a directory and its direct sub-directory are groups, then they belong to the same projects and a group in nested in the other
* if the parent directory of a group is not a group, then the group is the root of a project
* if a directory has a `src` sub-directory then it is a group
* a directory and a `src` sub-directory correspond to the same group

These rules are just crazy and historical and a better scheme is needed, that is KISS and POLA.

A proposal could be to add a specific metadata file in a directory to say that the directory is a project root directory.
So a module can find to which project if belong to

questions:

* how to name it? 
* what kind of metadata? version, author, copyright, compilation options?
* what format for the metadata? xml, json, ini, specific syntax, etc.?

A second need is for a project to find all its modules.
This is used for some commands that can take a project in order to process all its module (eg nitunit)
This will also be used to import modules from the project without having to thing about namespace and directories

Because of this, do we want that any cruft accessible from the project directory is considered a module of the project? Since when I program, I tend to let a lot of broken test-files in directories, my answer was no. There is also the case of generated files. Do we want that all the generated files `foo/.nitunit/test*.nit` belong to the project foo?

Currently, only subdirectory that are themselves groups belong to the same project, therefore this basically limit the project ownership to subdir that have an default module (even empty).
This is the explanation to the various empty `example.nit` files in `lib`. They make the `example` directory qualify as group and this belong to the same project that their parent directory.

For this part of the issue, the right specification seems to be something among:

* accept any source-file accessible from the project root directory
* like the previous but add optional include+exclude rules in the project metadata file
* use a specific group-metadata file to distinguish random-directory from groups and accept in the project only sub directories that are groups (and their subdirectories that are groups)

So, what do people think is the best trade-of as a developer (that use vim nevertheless)?

**Comment by R4PaSs** [comment]
I like the idea of having a specific syntax project descriptor.

Various metadata could be expressed like dependencies (and version of dependencies, maybe optional if we want the latest version ?), author, license, internationalization files, etc.
If we have our own format, we could produce easy to understand errors and it could be validated by a tool (once again `picknit` could do the job ?).
It can evolve depending of the future requirements easily by updating the grammar also.

Since we spoke once of a package manager for Nit akin to pip or gem (`picknit` ?), we could use a similar structure to those packages for our own.

Ruby gems are organized in folders, lib is the source code (I guess we could call it src, either would be ok from my point of view), tests contain the tests for the project and a binary folder receives the compiled version of the binaries.
Since we do not support shared libraries, the bin folder would not necessarily be available I guess.
But the [.rubyspec](http://guides.rubygems.org/specification-reference/) file might be something to inspire from.
They use standard Ruby to specify the contents of the project, we could do that, though I'm not sure how useful this would be since the informations are hardly useful at runtime.
The library's documentation could be distributed in a doc folder, all generated via nitdoc.
A final `README.md` would be the user's entry point to the library for basic information.

Python use .txt files in their eggs it would seem, I'm not that familiar with their distribution practices, but it is specified as such in this [document](https://pythonhosted.org/setuptools/formats.html#dependency-metadata).
I'm not necessarily convinced by the idea, though this might be relevant ? I don't know.

There is still the end-user's point of view.
For the import of a project in code, I guess the schema to go with could be akin to Java's.
Something along the lines of `project::(group::)*module?`, with intra-project imports not requiring the `project::` part ?

That's just the tip of the iceberg I guess, but if you guys have any more ideas, I'd love to hear more !

While we're on the naming side of things, why not introducing the .basket file type for Nit projects ? :)

**Comment by privat** [comment]
@R4PaSs ouch, I did not know that the current packaging specification in languages was this awful.
Can you 1. continue you overview for other languages (haskel/cabal, perl/cpan, go/?, etc)
2. collect the existing metadata that are considered (to have an idea of what we could have to support)

I do not understand the point about the .basket... Oh picknit basket. I got it now. So it means that when archiving a directory in a single files for easy distribution, like a rnr.basket.tar.gz, it is a basket ball?

Note: the import rule are not really the point of the present issue. I added a TL;DR to the original issue to keep the discussion focussed:)

**Comment by Morriar** [comment]
I used `npm` (node package manager) a lot and their way to [declare packages](https://docs.npmjs.com/files/package.json) is pretty easy and quick.

Everything is declared in a `package.json` file like this:

~~~json
{
  "name": "my-module",
  "version": "10.3.1",
  "description": "An example module to illustrate the usage of a package.json",
  "author": "Your Name <you.name@example.org>",
  "license": "MIT",
  "main": "lib/foo.js",
  "repository": {
    "type": "git",
    "url": "https://github.com/nodejitsu/browsenpm.org"
  },
  "bugs": {
    "url": "https://github.com/nodejitsu/browsenpm.org/issues"
  },
  "dependencies": {
    "async": "~0.8.0",
    "plates": "https://github.com/flatiron/plates/tarball/master"
  },
  "files": [
     "src/foo.nit"
     "src/bar/"
  ]
}
~~~

I think that the `files` keys answer your question.

In response to @R4PaSs, I don't like the idea of a specific DSL, there is a ****load of possible format to express what we want with a lot of already existing tools to writte/parseit. In fact, a simple ini file would suffice.

**Comment by privat** [comment]
@Morriar This is nice.
Maybe json is overkill for the specific kind of information needed (although it is coherent for nodejs) and a simple ini can be enough (with still having the possibility for some keys to refer to some other external files with a format adapted to the goal of these additional files).

I do not think that `files` answers directly my question about how do the loader can collect all the nit source files from the specified project. The options I see (but there might be other) are: a full recursive search (a la `find -name '*.nit'`) or specific rules to discover the directories (as currently). In fact the more I think about it the more I would prefer a full search since it is more pola , more kiss and that a specific directory structure does not seem to be able to solve a lot of issues in fact.

So, what about the following:
* a `project.ini` file in a directory indicates that this is the root directory of a nit project
* nit files are collected trough a recursive search. Maybe skip hidden files and directories, for instance avoiding things generated by nitunit. Maybe also skip subdirectories that contain a other `project.ini` and consider them as different unrelated projects (I am not sure I want to introduce crazy sub-project concepts)

For a module, to find its project could be just to `cd ..` until it finds a `project.ini` or the `/` directory. If / is found then it means that the module is standalone in its own personal project.

For the nitpath (-I/--include), nothing seems to be different than actually. except that a direct subdirectory without a `project.ini` file will not be considered as a potential nit project (or should it?)

For the content of `project.ini` I do think that a *convention over configuration* approach could be useful with sane default values for keys.

So, what about something like that?

~~~
[description]
name=foo # default is name of the directory?
author=John Doe <jdoe@example.com> # default is null
#no description because README.md?
homepage=http://example.com/foo
[source]
include=src/ # default is .
exclude=src/crap # default is null
[resources] # the following value are the default ones
po=po/ # for localization
assets=assets/ # for mnit or app.nit
~~~

In fact it seems that the ini might also be overkill since the sections are basically useless in this example. Maybe the example is not complex enough. Anyway, it is just random thought.

**Comment by Morriar** [comment]
@privat 

Ini is quite ok, but I don't see the point of sections if we don't have key collision, and for such of a configuration file, we shouldn't.

Plus, we already have a ini parser in lib.

The description can be useful for an hypothetical packet manager in which we don't want to implement a README.md parser just to collect the synopsis of the project.

**Comment by R4PaSs** [comment]
I don't think a project should contain any sub-project as you mentioned @privat.

They should be cleanly separated in different directories, the .ini would serve as a way to localise the different project repositories needed to build the current one.

The .ini files could also express dependencies on external resources like C libs or such.

Maybe something along these lines if we choose to go with a .ini file for description:

~~~ini
name=foo # default is name of the directory?
author=John Doe <jdoe@example.com> # default is null
#no description because README.md?
homepage=http://example.com/foo
include=src/ # default is .
exclude=src/crap # default is null
po=po/ # for localization
assets=assets/ # for mnit or app.nit
[dependencies]
[project]
name=standard
location= #empty until localised by a tool like picknit or via manual configuration
version=0.74+ #we can express bounds on version numbers ?
# something along the lines of + or a range if we need something that was available from versions x to y ?
[project]
name=socket
location= #to fill
version= 0.1.12 .. 0.2
[library]
name=boehmgc
location= /bin/
version= *
~~~

And I agree, the description of the project should be in the README.md file of the project, the .ini should only handle configuration purposes.

I don't know how useful the informations of repository and bugs are useful in a project ? Is it really relevant in a project description file ?

**Comment by Morriar** [comment]
@R4PaSs I don't think you can nest sections like that since the start of a new section automatically closes the previous one. This is why I proposed JSON!

**Comment by R4PaSs** [comment]
I was not aware of this limitation with .ini, in this case, JSON might make sense !
As long as we don't choose XML, I'm fine with it :)

**Comment by privat** [comment]
In fact, you can nest things in ini, this is used a lot, the idea is that to give to a section a full name

~~~ini
[foo]
a=1
[foo.bar]
b=2
~~~

Morever, we can also use sections to isolate specific configurations of various tools

~~~ini
# ...
[nitpretty]
line-width = 210
[nitdoc]
piwik-tracker = http://piwik.example.com
~~~
